common/rate-limiter/src/lib.rs

```diff
@@ -0,0 +1,485 @@
+        let tokens = self.tokens;
+        let tokens_left = tokens.saturating_sub(requested);
+        let tokens_allowed = min(tokens, requested);
+        self.tokens = tokens_left;
+
+        tokens_allowed
+    }
+
+    /// Tells us when the next refill will occur, when we don't need to worry about a whole batch
+    /// not being rate limited.  Useful where partial flows are okay.
+    pub fn time_until_next_refill(&self) -> Duration {
+        duration_saturating_sub(ONE_SEC, self.last_refresh_time.elapsed())
+    }
+
+    /// Tells us when an entire batch will make it through.  Useful for Async work to wait until
+    /// all tokens are ready.  Returns `None` if it is never possible.
+    pub fn time_until_tokens_needed(&self, requested: usize) -> Option<Duration> {
+        if self.open {
+            Some(ZERO_SEC)
+        } else if self.size < requested {
+            // This means the batch can never succeed
+            None
+        } else {
+            let tokens_needed = requested.saturating_sub(self.tokens);
+
+            let intervals = (tokens_needed as f64 / self.rate as f64).ceil() as u64;
+            Some(duration_saturating_sub(
+                Duration::from_secs(intervals),
+                self.last_refresh_time.elapsed(),
+            ))
+        }
+    }
+
+    /// Add new tokens, this can also be used to add unused tokens back
+    /// Ensures bucket doesn't overfill
+    pub fn add_tokens(&mut self, new_tokens: usize) {
+        self.tokens = min(self.size, self.tokens.saturating_add(new_tokens));
+    }
+}
+
+const ZERO_SEC: Duration = Duration::from_secs(0);
+const ONE_SEC: Duration = Duration::from_secs(1);
+
+/// Right now `Duration::saturating_sub` is unstable, so I've added this helper function for now
+fn duration_saturating_sub(duration1: Duration, duration2: Duration) -> Duration {
+    if duration1 <= duration2 {
+        ZERO_SEC
+    } else {
+        duration1 - duration2
+    }
+}
+
+/// A rate limiter for `AsyncRead` interfaces to rate limit read bytes
+///
+/// This will pause and wait to send any future bytes until it's permitted to in the future
+pub struct AsyncReadRateLimiter<T: AsyncRead + Unpin> {
+    reader: T,
+    rate_limiter: Arc<Mutex<Bucket>>,
+}
+
+impl<T: AsyncRead + Unpin> AsyncReadRateLimiter<T> {
+    pub fn new(reader: T, rate_limiter: Arc<Mutex<Bucket>>) -> Self {
+        Self {
+            reader,
+            rate_limiter,
+        }
+    }
+}
+
+impl<T: AsyncRead + Unpin> AsyncRead for AsyncReadRateLimiter<T> {
+    fn poll_read(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut [u8],
+    ) -> Poll<std::io::Result<usize>> {
+        let requested = buf.len();
+        let allowed = self.rate_limiter.lock().acquire_tokens(requested);
+
+        // Stop now if we're completely throttled
+        if allowed == 0 {
+            sleep(self.rate_limiter.lock().time_until_next_refill());
```

Comment:
don't want to use `thread::sleep` here, since it'll block the actual tokio runtime worker thread (which will block the entire Peer actor and prevent other futures from running on this thread; there are usually at most `num_cores` runtime worker threads). instead use the `tokio::time::Delay` future (created with `tokio::time::delay_for`).

unfortunately, we can't just await the delay here, since we're writing a poll function. when we write poll functions, it's more like writing a state machine. it might be helpful to think about the states and when you transition between them:

![Screen Shot 2020-12-16 at 10 31 30](https://user-images.githubusercontent.com/918989/102390919-de71c700-3f89-11eb-910f-7ea766df80d1.png)

you can encode the state like

```rust
use tokio::time::Delay;

struct AsyncReadRateLimiter {
    reader: T,
    rate_limiter: Arc<Mutex<Bucket>>,
    waiting: Option<Delay>,
}
```

you might need a `loop { .. }`

Reply:
I "think" I fixed this, there's now a test to check it out